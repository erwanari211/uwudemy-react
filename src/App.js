import React, { useState, useEffect } from 'react';
import './App.css';
import {
  BrowserRouter as Router,
  Switch,
  Route,
  useRouteMatch,
  useParams
} from "react-router-dom";

import { useDispatch, useSelector } from 'react-redux'

import ProductList from './components/ProductList'
import Home from './pages/Home'
import About from './pages/About'
import Navbar from './components/Navbar'
import Cart from './pages/Cart'
import ContactUs from './pages/ContactUs'

/**
 * 3. Function Component
 */
function App() {
  return (
    // 9. React Router
    <Router>
      {/* 5. Composing */}
      <Navbar />
      <div>
        <Switch>          
          {/* 5. Composing */}
          <Route path="/products">
            <Products />
          </Route>
          <Route path="/product-detail">
            <ProductDetail />
          </Route>
          <Route path="/cart">
            <Cart />
          </Route>
          <Route path="/about">
            <About />
          </Route>
          <Route path="/contact-us">
            <ContactUs />
          </Route>
          <Route path="/">
            <Home />
          </Route>
        </Switch>
      </div>
    </Router>
  );
}

/**
 * 3. Function Component
 * 6. Extracting Component
 */
function Products() {
  // 10. State Management Redux
  const products = useSelector(state => state.products)
  const dispatch = useDispatch()

  // 4. LifeCycle Method
  // 8. Hooks
  useEffect(() => {
    console.log('Display products')
  }, [])

  const handleSearchProducts = (e) => {
    e.preventDefault()
    let search = inputSearch.current.value
    // 10. State Management Redux
    dispatch({ type: 'FILTER PRODUCT', search: search })
  }

  const inputSearch = React.createRef()
  

  return (
    <div className="container">
      <h2>Products</h2>

      <div className="row mb-4">
        <div className="col-md-4">
          <form>
            <div className="form-group">
              <input className="form-control" name="search" placeholder="Search" ref={inputSearch} />
            </div>
            
            <button className="btn btn-primary" onClick={handleSearchProducts}>Search</button>
          </form>
        </div>
      </div>

      {/* 5. Composing */}
      <ProductList products={products} />
    </div>
  );
}

/**
 * 3. Function Component
 * 6. Extracting Component
 */
function ProductDetail(){
  let match = useRouteMatch();

  // 10. State Management Redux
  const dispatch = useDispatch()

  // 7. Lifting State Up
  const addToCart = (product) => {
    console.log(product)
    // 10. State Management Redux
    dispatch({ type: 'ADD TO CART', product: product })
  }

  return (
    <div>
      <Switch>
        <Route path={`${match.path}/:productId`}>
          {/* 5. Composing */}
          <Product addToCart={addToCart} />
        </Route>
        <Route path={match.path}>
          404
        </Route>
      </Switch>
    </div>
  );
}

/**
 * 3. Function Component
 * 6. Extracting Component
 */
function Product(props) {
  let { productId } = useParams();

  // 10. State Management Redux
  const products = useSelector(state => state.products)

  // 8. Hooks
  const [product, setProduct] = useState({})
  const [productExist, setProductExist] = useState(false)

  // 8. Hooks
  // 4. LifeCycle Method
  useEffect(() => {
    let item = products.filter((item) => parseInt(item.id) === parseInt(productId) );
    if(item.length === 1){
      setProduct(item[0])
      setProductExist(true)
    }
    console.log(item)
  }, [])

  // 7. Lifting State Up
  const handleAddToCart = () => {
    props.addToCart(product)
  }

  if (productExist) {
    return (
      <div className="container">
        <div className="row">
          <div className="col-md-12">
            <h2>Detail Product</h2>
            <div className="card mb-4">
              <div className="card-body">
                <div className="media">
                  <img className="mr-3" src={product.img} alt={product.name} />
                  <div className="media-body">
                    <h5 className="mt-0">{product.name}</h5>
                    <p className="text-success">US${product.price}</p>

                    {/* 12. Lists */}
                    <ul className="list-inline">
                      {product.tags.map((tag, index) => {
                        return (
                          <li className="list-inline-item" key={index}>
                            <span className="badge badge-info">{tag}</span>
                          </li>                        
                        )
                      })}
                    </ul>  
                    <button className="btn btn-primary" onClick={handleAddToCart}>Add to cart</button>
                  </div>
                </div>
              </div>              
            </div>
          </div>
        </div>
      </div>
    )
  }

  return (
    <div className="container">
      <div className="row">
        <div className="col-md-12">
          <h2>Detail Product</h2>
          <div className="card mb-4">
            <div className="card-body">
              Product not found
            </div>            
          </div>
        </div>
      </div>
    </div>
  );
}


export default App;
