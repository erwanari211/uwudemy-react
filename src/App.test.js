import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';

import App from './App';

/**
 * 13. unit testing
 */
// test('renders learn react link', () => {
//   const { getByText } = render(<App />);
//   const linkElement = getByText(/learn react/i);
//   expect(linkElement).toBeInTheDocument();
// });

test('Test Snapshot', () => {
  const component = renderer.create(<App />);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Find text uwudemy', () => {
  const { queryAllByText } = render(<App />);

  expect(queryAllByText(/UWUdemy/i)).toBeTruthy();
});

test('Find text lihat semua kelas', () => {
  const { queryAllByText } = render(<App />);

  expect(queryAllByText(/Lihat Semua Kelas/i)).toBeTruthy();
});
