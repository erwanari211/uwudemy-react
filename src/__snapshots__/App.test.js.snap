// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Snapshot 1`] = `
Array [
  <div
    className=""
  >
    <nav
      className="navbar navbar-expand-sm navbar-light bg-light"
    >
      <a
        className="navbar-brand"
        href="/"
        onClick={[Function]}
      >
        UWUdemy
      </a>
      <button
        aria-controls="navbarSupportedContent"
        aria-expanded="false"
        aria-label="Toggle navigation"
        className="navbar-toggler"
        data-target="#navbarSupportedContent"
        data-toggle="collapse"
        type="button"
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        className="collapse navbar-collapse"
        id="navbarSupportedContent"
      >
        <ul
          className="navbar-nav mr-auto"
        >
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="/"
              onClick={[Function]}
            >
              Home 
              <span
                className="sr-only"
              >
                (current)
              </span>
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              data-testid="link to product"
              href="/products"
              onClick={[Function]}
            >
              Products
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="/cart"
              onClick={[Function]}
            >
              Cart
            </a>
          </li>
        </ul>
        <ul
          className="navbar-nav"
        >
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="/about"
              onClick={[Function]}
            >
              About
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="/contact-us"
              onClick={[Function]}
            >
              Contact Us
            </a>
          </li>
        </ul>
      </div>
    </nav>
  </div>,
  <div>
    <div
      className="container"
      style={
        Object {
          "paddingBottom": "50px",
        }
      }
    >
      <h2>
        Home
      </h2>
      <div
        className="jumbotron"
      >
        <h1
          className="display-4"
        >
          UWUdemy
        </h1>
        <p
          className="lead"
        >
          Kursus Online. Belajar Koding
        </p>
        <hr
          className="my-4"
        />
        <p>
          Pelajari topik apa saja, kapan saja. Jelajahi ribuan kursus masing-masing mulai US$12,99.
        </p>
        <p
          className="lead"
        >
          <a
            className="btn btn-primary btn-lg"
            href="/products"
            onClick={[Function]}
          >
            Lihat Semua Kelas
          </a>
        </p>
      </div>
      <div
        className="row"
      >
        <div
          className="col-md-4"
        >
          <strong>
            100.000 kursus online
          </strong>
          <br />
          <span>
            Nikmati berbagai topik baru
          </span>
        </div>
        <div
          className="col-md-4"
        >
          <strong>
            Instruksi dari pakarnya
          </strong>
          <br />
          <span>
            Temukan instruktur yang tepat untuk Anda
          </span>
        </div>
        <div
          className="col-md-4"
        >
          <strong>
            Akses seumur hidup
          </strong>
          <br />
          <span>
            Belajar sesuai jadwal Anda
          </span>
        </div>
      </div>
    </div>
  </div>,
]
`;
